### PR TITLE
Storing strings in JSONField.

### PR DESCRIPTION
**Pull request by arnau126**
Be able to store strings in JSONField (this feature was broken).

Also copy code from 'django.contrib.postgres.forms.jsonb.JSONField' in order to fix some issues related with redisplaying wrong user input.

This PR is motivated by: #349 (which now it is not longer necessary).

Checklist:

- [x] Line added to HISTORY.rst, or N/A

Test Plan:
- Create some test about serializing strings.
- Copy two tests from ' django.tests.postgres_tests.test_json.py'.
- All current and new tests pass.

### CODE REVIEW

**Comment by arnau126** [comment]
TODO: tests don't pass in `django1.8` because in this version `Field` doesn't have the `disabled` attribute, yet.

**Comment by arnau126** [comment]
any thoughts, @adamchainz ?

**Comment by sergiomb2** [comment]
@arnau126 	Merge branch 'master' into string_jsons 

instead merge , you use rebase , git rebase upstream/master && git push --force (I think you need force the push)

**Comment by adamchainz** [comment]
@arnau126 don't do anything I'll rebase the PR

I have been a bit busy recently but will look at this soon.

**Comment by arnau126** [comment]
@sergiomb2 Thanks for the advice.

@adamchainz Okay. Let me know if any change is needed.

**Comment by arnau126** [comment]
@adamchainz Sorry for the rush, but have you been able to take a look at this?
Besides, when do you expect to release? I would like to use JSONSet in the project I'm currently working on.

**Comment by adamchainz** [comment]
Looking at it now finally.

I am going to merge this as two commits, one for the form field and one for the model field. I think the model field fix could break some people's applications if they've been doing e.g. `model.attrs = json.dumps({'a': 'dict'})`, so I think it should go in a "breaking change" release, just in case.

File: HISTORY.rst [hunk 1]
```diff
@@ -7,6 +7,8 @@ Pending
 -------
 
 * (Insert new release notes below this line)
+* Fixed ``JSONField`` to allow storing strings as top-level json objects.
+* Fixed ``JSONField`` so that it doesn't overquote inputs when redisplaying the form due to invalid user input.
```
Comment (adamchainz) [inline HISTORY.rst]: this file is wrapped at 80 chars

File: django_mysql/forms.py [hunk 2]
```diff
@@ -240,8 +240,6 @@ class JSONField(forms.CharField):
     widget = forms.Textarea
 
     def to_python(self, value):
-        if self.disabled:
```
Comment (adamchainz) [inline django_mysql/forms.py]: removing this logic isn't the right approach, we want to work with disabled on Django 1.9+. instead we should do:

```
if getattr(self, 'disabled', False):
```

or

```
if django.VERSION[:2] >= (1, 9) and self.disabled:
```

Allowing the code to work on both versions. And add a test too, to get coverage of the lines.

File: tests/testapp/test_forms.py [hunk 3]
```diff
@@ -306,4 +307,37 @@ def test_invalid(self):
...
+        When displaying a bound form (typically due to invalid input), the form
+        should not overquote JSONField inputs.
+        """
+        class JsonForm(forms.Form):
+            name = forms.CharField(max_length=2)
+            jfield = JSONField()
+
+        # JSONField input is fine, name is too long
+        form = JsonForm({'name': 'xyz', 'jfield': '["foo"]'})
+        self.assertIn('[&quot;foo&quot;]</textarea>', form.as_p())
```
Comment (adamchainz) [inline tests/testapp/test_forms.py]: our tests are on pytest and use plain `assert` everywhere, so this shoudl be `assert '<snip>' in form.as_p()` etc.

**Comment by adamchainz** [comment]
Merged form field change in 9930b2009c72cea7d2bd15c474e0edd209e5138a with some touch-up in 556d8987845fb3a11f73a2a86855fd311126967f

**Comment by arnau126** [comment]
Okay!
And thanks for all your remarks.

**Comment by kyteague** [comment]
FYI, the bug that this PR addresses caused us two production outages because bad JSON data was saved back to the database (our application was expecting a JSON list not a string). A quick merge would be appreciated, thanks!

**Comment by arnau126** [comment]
This bug is already merged and released in 1.2.0:
https://github.com/adamchainz/django-mysql/blob/master/HISTORY.rst#120-2017-05-14

adamchainz said:
> I am going to merge this as two commits, one for the form field and one for the model field.

> Merged the form field stuff in 2f03d96 , fixed up in 504fd9c. Thanks!

**Comment by adamchainz** [comment]
I think @kyteague wants the model field fix. I'm going to merge it today

**Comment by adamchainz** [comment]
Merged the model field fix in 961b9f5c817a599e3a6f4a033a63d7e5cf326213, touched up in 12616152e5129559daacba8957cb738319eee586. Thanks!

**Pull request closed without merging**